File: algorithms/2053.kth-distinct-string-in-an-array/solution.rb
# @param {String[]} arr
# @param {Integer} k
# @return {String}
def kth_distinct(arr, k)
  h = Hash.new([0, 0])

  arr.each_with_index do |c, i|
    if h.has_key?(c)
      h[c] = [h[c][0] + 1, i]
    else
      h[c] = [1, i]
    end
  end

  counts = h.select { |_, v| v[0] == 1 }.sort_by { |_, v| v[1] }
  counts.size >= k ? counts[k - 1][0] : ""
end

File: algorithms/2053.kth-distinct-string-in-an-array/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn kth_distinct(arr: Vec<String>, k: i32) -> String {
        let mut hm: HashMap<String, Vec<i32>> = HashMap::new();

        for (i, s) in arr.into_iter().enumerate() {
            let entry = hm.entry(s).or_insert(vec![0, i as i32]);
            (*entry)[0] += 1;
        }

        let mut counts: Vec<(&String, i32)> = hm
            .iter()
            .filter(|(_, v)| v[0] == 1)
            .map(|(k, v)| (k, v[1]))
            .collect();
        counts.sort_by(|a, b| a.1.cmp(&b.1));

        if counts.len() >= k as usize {
            counts[k as usize - 1].0.to_string()
        } else {
            "".to_string()
        }
    }
}
